### PR TITLE
[FLINK-3943] Add support for EXCEPT operator

### PR DESCRIPTION
**Pull request by mushketyk**
Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.

- [x] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)

- [x] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added

- [x] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

### CODE REVIEW

**Comment by mushketyk** [comment]
Renamed UnionITCase to SetOperationsITCase as suggested here: https://github.com/apache/flink/pull/2159

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetMinus.scala [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+        rightDataSet =
+          right.asInstanceOf[DataSetRel].translateToPlan(tableEnv, Some(leftDataSet.getType))
+      case _ =>
+        leftDataSet = left.asInstanceOf[DataSetRel].translateToPlan(tableEnv, expectedType)
+        rightDataSet = right.asInstanceOf[DataSetRel].translateToPlan(tableEnv, expectedType)
+    }
+
+    val minusRes = leftDataSet.minus(rightDataSet)
+    if (!all) {
+      minusRes.distinct()
```
Comment (wuchong) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetMinus.scala]: I think it's better to remove duplicate records in CoGroup, emit only one record instead of every record from left data set when it is a minus without all. There is no need a distinct afterwards. It's the more robust choice because it won't create a huge intermediate result in case of many duplicate records.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/operators.scala [hunk 2]
```diff
@@ -236,6 +236,32 @@ case class Aggregate(
...
+  }
+
+  override def validate(tableEnv: TableEnvironment): LogicalNode = {
+    val resolvedMinus = super.validate(tableEnv).asInstanceOf[SetMinus]
+    if (left.output.length != right.output.length) {
+      failValidation(s"Set minus two table of different column sizes:" +
+        s" ${left.output.size} and ${right.output.size}")
+    }
+    val sameSchema = left.output.zip(right.output).forall { case (l, r) =>
+      l.resultType == r.resultType && l.name == r.name }
```
Comment (wuchong) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/operators.scala]: Set minus do not need the field names exactly same. So I suggest to remove the name constraint. Actually, it is the same for Union. However, for Union,  the problem is so complex that can't be simply resolved.

File: docs/apis/table.md [hunk 3]
```diff
@@ -695,6 +718,30 @@ val result = left.unionAll(right);
...
+{% highlight scala %}
+val left = ds1.toTable(tableEnv, 'a, 'b, 'c);
+val right = ds2.toTable(tableEnv, 'a, 'b, 'c);
+val result = left.minus(right);
+{% endhighlight %}
+      </td>
+    </tr>
+
+    <tr>
+      <td><strong>UnionAll</strong></td>
```
Comment (wuchong) [inline docs/apis/table.md]: UnionAll  -> MinusAll

File: docs/apis/table.md [hunk 4]
```diff
@@ -873,7 +920,7 @@ val result = tableEnv.sql(
 
 #### Limitations
 
-The current version of streaming SQL only supports `SELECT`, `FROM`, `WHERE`, and `UNION` clauses. Aggregations or joins are not supported yet.
+The current version of streaming SQL only supports `SELECT`, `FROM`, `WHERE`, `UNION` and `EXCEPT` clauses. Aggregations or joins are not supported yet.
```
Comment (wuchong) [inline docs/apis/table.md]: It seems that Stream SQL hasn't  supported `EXCEPT`.

**Comment by wuchong** [comment]
Hi @mushketyk, I think we should remove duplicate records in CoGroup instead of using `distinct`. Others looks good to me.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/operators.scala (hunk 2 above)
Comment (mushketyk) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/operators.scala]: Hm... I don't think I fully understand reasoning behind this. Could you elaborate please?

Comment (wuchong) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/operators.scala]: I think both tables must have the same number of fields with similar data types, but they can accept different field names. e.g.  `testMinusDifferentFieldNames`  should pass without no exception. Correct me if I'm wrong.

Comment (mushketyk) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/operators.scala]: I don't think it makes much sense to perform minus between sets {(1, 2, 3), (4, 5, 6)} and {(1, 2, 3, 10, 20), (4, 5, 6, 50, 60)}. Such operation would likely be a sign of a user error and it seems to me that it's better to throw an exception than silently do something surprising. Therefore I don't think that testMinusDifferentFieldsNames should pass.

As for a case when we have different names but identical field count and type I don't see any difference between a union and minus. It seems to me that both union and minus are just set operations on tuples and technically both can work in case when we have tables with different field names.

@fhueske What do you think about it?

File: docs/apis/table.md (hunk 3 above)
Comment (mushketyk) [inline docs/apis/table.md]: Good catch. Will change this.

File: docs/apis/table.md (hunk 4 above)
Comment (mushketyk) [inline docs/apis/table.md]: Accidentally changed this. Will revert this edit.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetMinus.scala (hunk 1 above)
Comment (mushketyk) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetMinus.scala]: Makes sense. Will update.

Comment (mushketyk) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetMinus.scala]: On the second hand "union" method in DataSet.java preserves duplicates and used to implement UNION ALL. UNION is implemented by using "distinct" operation on top of UNION ALL.
It seems that what you suggest will only add code duplication since it will basically implement a specialized version of "distinct" operator specifically for the "minus" method.

I have no strong preferences though.  @fhueske what do you think?

**Comment by mushketyk** [comment]
Updated documentation as suggested.

File: flink-java/src/main/java/org/apache/flink/api/java/DataSet.java [hunk 5]
```diff
@@ -1193,21 +1196,46 @@ public long count() throws Exception {
...
 		return new UnionOperator<>(this, other, Utils.getCallLocationName());
 	}
 
+	/**
+	* Creates a set minus of this DataSet with an other DataSet. The other DataSet must be of the same data type.
+	*
+	* @param other The other DataSet which is set minus with the current DataSet.
+	* @return The resulting DataSet.
+	*/
+	public CoGroupOperator<T, T, T> minus(DataSet<T> other){
```
Comment (fhueske) [inline flink-java/src/main/java/org/apache/flink/api/java/DataSet.java]: This issue is about adding `EXCEPT` to the Table. The DataSet API which is touched here is a rather low level API and we are quite careful about adding new operators. Therefore, changes to the DataSet API should go through a separate JIRA issue. Please move this code to `DataSetMinus` class and revert the changes to this file. You can open a JIRA issue to discuss adding a `minus` operator to the DataSet API. Thank you.

File: docs/apis/table.md [hunk 6]
```diff
@@ -536,6 +536,29 @@ Table result = left.unionAll(right);
     </tr>
 
     <tr>
+      <td><strong>Minus</strong></td>
+      <td>
+        <p>Similar to a SQL EXCEPT clause. Returns elements from the first table that do not exist in the second table. Both tables must have identical schema, i.e., field names and types.</p>
```
Comment (fhueske) [inline docs/apis/table.md]: Please rephrase to "Except returns records from the first table that do not exist in the second table. Records that exist multiple times in the first table are returned exactly once, i.e., duplicates are removed.".

File: docs/apis/table.md [hunk 7]
```diff
@@ -536,6 +536,29 @@ Table result = left.unionAll(right);
...
+Table right = tableEnv.fromDataSet(ds2, "a, b, c");
+Table result = left.minus(right);
+{% endhighlight %}
+      </td>
+    </tr>
+
+    <tr>
+      <td><strong>MinusAll</strong></td>
+      <td>
+        <p>Similar to a SQL EXCEPT ALL clause. Returns elements from the first table that do not exist in the second table without removing duplicates. Both tables must have identical schema, i.e., field names and types.</p>
```
Comment (fhueske) [inline docs/apis/table.md]: Please rephrase to "Except All returns the records that do not exist in the second table. A record that is present n times in the first table and m times in the second table is returned (n - m) times, i.e., as many duplicates as are present in the second table are removed.".

File: docs/apis/table.md [hunk 8]
```diff
@@ -873,7 +920,7 @@ val result = tableEnv.sql(
 
 #### Limitations
 
-The current version of streaming SQL only supports `SELECT`, `FROM`, `WHERE`, and `UNION` clauses. Aggregations or joins are not supported yet.
+The current version of streaming SQL only supports `SELECT`, `FROM`, `WHERE` and `UNION` clauses. Aggregations or joins are not supported yet.
```
Comment (fhueske) [inline docs/apis/table.md]: We try to consistently use [serial commas](https://en.wikipedia.org/wiki/Serial_comma) in this document.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/operators.scala (hunk 2 above)
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/operators.scala]: I think @wuchong refers exactly to the last case. If number of fields and field types are identical, it should be possible to do a `EXCEPT` even if the field names are not the same.
The situation for `UNION` is a bit special due to Flink's internals and the way the `RowTypeInfo` is implemented but I think we can remove that restriction in the future. 

So, we must keep the checks for number of fields and field types but can remove the check of the field names, IMO.

File: flink-java/src/main/java/org/apache/flink/api/java/DataSet.java [hunk 9]
```diff
@@ -1193,21 +1196,46 @@ public long count() throws Exception {
...
+		return coGroup(other)
+			.where("*")
+			.equalTo("*")
+			.with(new RichCoGroupFunction<T, T, T>() {
```
Comment (fhueske) [inline flink-java/src/main/java/org/apache/flink/api/java/DataSet.java]: A non-rich `CoGroupFunction` is sufficient.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetMinus.scala (hunk 1 above)
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetMinus.scala]: I would like to move the code of `DataSet.minus()` here. However, I think the semantics of `EXCEPT ALL` are a bit different than in your implementation. It is not simply checking if there is a match in the second input and forwarding everything if there is none. It basically removes for each match in the second input one matching record from the first input (see also the [PostgreSQL docs](https://www.postgresql.org/docs/9.4/static/sql-select.html#SQL-EXCEPT)).

I would be in favor of an implementation that is similar to @wuchong's implementation of `INTERSECT` / `INTERSECT ALL` in PR #2159.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/rules/dataSet/DataSetMinusRule.scala [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+    classOf[LogicalMinus],
+    Convention.NONE,
+    DataSetConvention.INSTANCE,
+    "DataSetMinusRule")
+{
+
+  /**
+    * Translate EXCEPT and EXCEPT ALL.
+    */
+  override def matches(call: RelOptRuleCall): Boolean = {
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/rules/dataSet/DataSetMinusRule.scala]: This method can be removed. It always returns `true` which is the default implementation which is overwritten here.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/table.scala [hunk 11]
```diff
@@ -426,6 +426,49 @@ class Table(
   }
 
   /**
+    * Set minus between two [[Table]]s. Similar to an SQL EXCEPT ALL.
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/table.scala]: Please describe the semantics of minus all in more detail.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/table.scala [hunk 12]
```diff
@@ -426,6 +426,49 @@ class Table(
...
+  def minusAll(right: Table): Table = {
+    // check that right table belongs to the same TableEnvironment
+    if (right.tableEnv != this.tableEnv) {
+      throw new ValidationException("Only tables from the same TableEnvironment can be unioned.")
+    }
+    new Table(tableEnv, SetMinus(logicalPlan, right.logicalPlan, true).validate(tableEnv))
+  }
+
+  /**
+    * Perform set minus between [[Table]]s with duplicate records removed.
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/table.scala]: Please describe the semantics of minus all in more detail:
- records of the first table are returned
- how many identical records are returned

File: flink-scala/src/main/scala/org/apache/flink/api/scala/DataSet.scala [hunk 13]
```diff
@@ -1353,6 +1353,16 @@ class DataSet[T: ClassTag](set: JavaDataSet[T]) {
...
 
   // --------------------------------------------------------------------------------------------
+  //  Minus
+  // --------------------------------------------------------------------------------------------
+
+  /**
+    * Creates a new DataSet containing the elements from `this` DataSet minus elements from `other`
+    * DataSet.
+    */
+  def minus(other: DataSet[T]): DataSet[T] = wrap(javaSet.minus(other.javaSet))
```
Comment (fhueske) [inline flink-scala/src/main/scala/org/apache/flink/api/scala/DataSet.scala]: Please remove these changes as well. Thanks

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/api/scala/batch/sql/SetOperationsITCase.scala [hunk 14]
```diff
@@ -121,4 +121,66 @@ class UnionITCase(
...
+
+    val result = tEnv.sql(sqlQuery)
+
+    val expected = "Hello\n" + "Hello world\n"
+    val results = result.toDataSet[Row].collect()
+    TestBaseUtils.compareResultAsText(results.asJava, expected)
+  }
+
+  @Test
+  def testExceptWithFilter(): Unit = {
```
Comment (fhueske) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/api/scala/batch/sql/SetOperationsITCase.scala]: Please remove this test. We need to be very careful about the build time of the project (we are very close to hit the 2h build timeout of Travis) and integration tests are quite time consuming. So we try to avoid integration tests that do not add to the test coverage.

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/api/scala/batch/sql/SetOperationsITCase.scala [hunk 15]
```diff
@@ -121,4 +121,66 @@ class UnionITCase(
...
+
+    val result = tEnv.sql(sqlQuery)
+
+    val expected = "Hi\n"
+    val results = result.toDataSet[Row].collect()
+    TestBaseUtils.compareResultAsText(results.asJava, expected)
+  }
+
+  @Test
+  def testExceptWithAggregation(): Unit = {
```
Comment (fhueske) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/api/scala/batch/sql/SetOperationsITCase.scala]: Please remove this test.

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/api/scala/batch/table/SetOperationsITCase.scala [hunk 16]
```diff
@@ -61,14 +62,28 @@ class UnionITCase(
     val ds1 = CollectionDataSets.getSmall3TupleDataSet(env).toTable(tEnv, 'a, 'b, 'c)
     val ds2 = CollectionDataSets.getSmall3TupleDataSet(env).toTable(tEnv, 'a, 'b, 'c)
 
-    val unionDs = ds1.union(ds2).select('c)
+    val minusDs = ds1.union(ds2).select('c)
```
Comment (fhueske) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/api/scala/batch/table/SetOperationsITCase.scala]: Why this change?

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/api/scala/batch/table/SetOperationsITCase.scala [hunk 17]
```diff
@@ -61,14 +62,28 @@ class UnionITCase(
...
 
-    val results = unionDs.toDataSet[Row].collect()
+    val results = minusDs.toDataSet[Row].collect()
     val expected = "Hi\n" + "Hello\n" + "Hello world\n"
     TestBaseUtils.compareResultAsText(results.asJava, expected)
   }
 
   @Test
+  def testSetMinusAllSameSets(): Unit = {
```
Comment (fhueske) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/api/scala/batch/table/SetOperationsITCase.scala]: Please remove this test.

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/api/scala/batch/table/SetOperationsITCase.scala [hunk 18]
```diff
@@ -139,4 +154,105 @@ class UnionITCase(
     // Must fail. Tables are bound to different TableEnvironments.
     ds1.unionAll(ds2).select('c)
   }
+
+  @Test
+  def testSetMinusAll(): Unit = {
```
Comment (fhueske) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/api/scala/batch/table/SetOperationsITCase.scala]: Can you combine this and the next test by using test data that covers both cases for different records, i.e., have some records with duplicates in this first, second, none, and both data sets.

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/api/scala/batch/table/SetOperationsITCase.scala [hunk 19]
```diff
@@ -139,4 +154,105 @@ class UnionITCase(
...
+    val minusDs = ds1.unionAll(ds2).minusAll(ds3).select('c)
+
+    val results = minusDs.toDataSet[Row].collect()
+    val expected = "Hello\n" + "Hello world\n" +
+      "Hello\n" + "Hello world\n"
+    TestBaseUtils.compareResultAsText(results.asJava, expected)
+  }
+
+  @Test
+  def testSetMinus(): Unit = {
```
Comment (fhueske) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/api/scala/batch/table/SetOperationsITCase.scala]: Can you combine this and the next test by using test data that covers both cases for different records, i.e., have some records with duplicates in this first, second, none, and both data sets.

**Comment by fhueske** [comment]
Hi @mushketyk, thanks for the PR! I added a few comments inline.

Best, Fabian
@wuchong, thanks for reviewing!

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/operators.scala (hunk 2 above)
Comment (wuchong) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/operators.scala]: Yes, I refer to the last case. I agree with @fhueske 's opinion, we can remove the check of field names in `EXCEPT` and `INTERSECT` now, and remove the restriction in `UNION` in the future.

**Comment by mushketyk** [comment]
@fhueske Thank you for the detailed review! I've updated my code according to your comments.

I noticed that @wuchong is performing some type conversions in his INTERSECT implementation: https://github.com/apache/flink/pull/2159/files#diff-a6c2112ca46d26fcf49f1edba1c73f75R121

Should I do something similar in the EXCEPT case? If yes, does it mean that my test coverage is not sufficient and does not cover some particular case?

**Comment by mushketyk** [comment]
@fhueske Hey. Sorry for bothering you. Do I need to change something else in this PR?

File: docs/apis/table.md [hunk 20]
```diff
@@ -536,6 +536,29 @@ Table result = left.unionAll(right);
     </tr>
 
     <tr>
+      <td><strong>Minus</strong></td>
+      <td>
+        <p>Similar to a SQL EXCEPT clause. Except returns records from the first table that do not exist in the second table. Duplicate records in the first table are returned exactly once, i.e., duplicates are removed. Both tables must have identical schema, i.e., field names and types.</p>
```
Comment (twalthr) [inline docs/apis/table.md]: "Except returns records" should be "Minus returns records" to be consistent. I would also use "left/right" table instead of first and second according to your example code.

File: docs/apis/table.md [hunk 21]
```diff
@@ -695,6 +718,30 @@ val result = left.unionAll(right);
     </tr>
 
     <tr>
+      <td><strong>Minus</strong></td>
+      <td>
+        <p>Similar to a SQL EXCEPT clause. Returns elements from the first table that do not exist in the second table. Both tables must have identical schema(field names and types).</p>
```
Comment (twalthr) [inline docs/apis/table.md]: Can you copy the description from the Java part also in here?

File: flink-java/src/main/java/org/apache/flink/api/java/DataSet.java [hunk 22]
```diff
@@ -18,8 +18,8 @@
 
 package org.apache.flink.api.java;
```
Comment (twalthr) [inline flink-java/src/main/java/org/apache/flink/api/java/DataSet.java]: I think you did not intend to change this file. Can you remove it from this PR?

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/operators.scala [hunk 23]
```diff
@@ -236,6 +236,32 @@ case class Aggregate(
   }
 }
 
+case class SetMinus(left: LogicalNode, right: LogicalNode, all: Boolean) extends BinaryNode {
```
Comment (twalthr) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/operators.scala]: I would still call it `Minus` to be in sync with `Union`, `Intersect`, etc. but that's just my personal opinion.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetMinus.scala [hunk 24]
```diff
@@ -0,0 +1,150 @@
...
+import org.apache.flink.api.common.typeutils.CompositeType
+import org.apache.flink.api.java.DataSet
+import org.apache.flink.api.table.BatchTableEnvironment
+import org.apache.flink.util.Collector
+
+import scala.collection.JavaConverters._
+import scala.collection.JavaConversions._
+
+/**
+  * Flink RelNode which matches along with DataSetOperator.
```
Comment (twalthr) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetMinus.scala]: This description does not make much sense.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetMinus.scala [hunk 25]
```diff
@@ -0,0 +1,150 @@
...
+  }
+
+  override def computeSelfCost (planner: RelOptPlanner, metadata: RelMetadataQuery): RelOptCost = {
+
+    val children = this.getInputs
+    val rowCnt = children.foldLeft(0D) { (rows, child) =>
+      rows + metadata.getRowCount(child)
+    }
+
+    planner.getCostFactory.makeCost(rowCnt, 0, 0)
```
Comment (twalthr) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetMinus.scala]: Shouldn't the cost be similar to an Intersect because both are using a CoGroup?

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetMinus.scala [hunk 26]
```diff
@@ -0,0 +1,150 @@
...
+        rightDataSet = right.asInstanceOf[DataSetRel].translateToPlan(tableEnv, expectedType)
+    }
+
+    val leftType = leftDataSet.getType
+    val rightType = rightDataSet.getType
+    val coGroupedDs = leftDataSet.coGroup(rightDataSet)
+
+    // If it is atomic type, the field expression need to be "*".
+    // Otherwise, we use int-based field position keys
+    val coGroupedPredicateDs =
```
Comment (twalthr) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetMinus.scala]: You can use "*" for all types. So the case distinction is not necessary.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetMinus.scala [hunk 27]
```diff
@@ -0,0 +1,150 @@
...
+      .name(s"intersect: $setMinusSelectionToString")
+  }
+
+  private def setMinusSelectionToString: String = {
+    rowType.getFieldNames.asScala.toList.mkString(", ")
+  }
+
+}
+
+class MinusCoGroupFunction[T](all: Boolean) extends CoGroupFunction[T, T, T] {
```
Comment (twalthr) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetMinus.scala]: This belongs into the table runtime package.

**Comment by twalthr** [comment]
@mushketyk Thanks for the PR.
I also reviewed the current status. I think it's good to compare your code with #2159 before you rework it.

**Comment by mushketyk** [comment]
@twalthr Thank you for your review! I've update the PR according to your comments.

**Comment by mushketyk** [comment]
@twalthr Updated PR according to your commits and rebased on top of the master branch to avoid merge conflicts.

**Comment by twalthr** [comment]
Thanks for updating the PR. I reviewed the code again. There were still some issues regarding the documentation and also `expectedType` handling in `DataSetMinus`, I fixed them myself and will merge now...

**Comment by mushketyk** [comment]
@twalthr Thank you for accepting my changes!
